### PR TITLE
Added LP pooling  layer

### PR DESCRIPTION
**Pull request by abh2k**
Added LP                                                     pooling layer. 
https://pytorch.org/docs/stable/generated/torch.nn.LPPool2d.html#torch.nn.LPPool2d

### CODE REVIEW

File: src/mlpack/methods/ann/layer/lp_pooling.hpp [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+   * @param strideWidth Width of the stride operation.
+   * @param strideHeight Width of the stride operation.
+   * @param floor Set to true to use floor method.
+   */
+  LpPooling(const size_t norm_type,
+              const size_t kernelWidth,
+              const size_t kernelHeight,
+              const size_t strideWidth = 1,
+              const size_t strideHeight = 1,
+              const bool floor = true);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/lp_pooling.hpp:52]: The parameter should align with the first one.

File: src/mlpack/methods/ann/layer/lp_pooling.hpp [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    const size_t cStep = input.n_cols / error.n_cols - offset;
+
+    arma::Mat<eT> unpooledError;
+    for (size_t j = 0; j < input.n_cols - cStep; j += cStep)
+    {
+      for (size_t i = 0; i < input.n_rows - rStep; i += rStep)
+      {
+        const arma::Mat<eT>& inputArea = input(arma::span(i, i + rStep - 1),
+            arma::span(j, j + cStep - 1));
+        size_t sum = arma::pow(arma::accu(arma::pow(inputArea, norm_type)), (norm_type-1) / norm_type);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/lp_pooling.hpp:204]: Looks like this exceeds the 80 line character limit.

File: src/mlpack/methods/ann/layer/lp_pooling.hpp [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+  //! Get the value of the rounding operation
+  bool const& Floor() const { return floor; }
+  //! Modify the value of the rounding operation
+  bool& Floor() { return floor; }
+
+  //! Get the value of the deterministic parameter.
+  bool Deterministic() const { return deterministic; }
+  //! Modify the value of the deterministic parameter.
+  bool& Deterministic() { return deterministic; }
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/lp_pooling.hpp:147]: I think the `deterministic` parameter isn't needed for the layer, as the functionality is the same for training and testing.

File: src/mlpack/tests/ann_layer_test.cpp [hunk 4]
```diff
@@ -3838,6 +3838,54 @@ TEST_CASE("TransposedConvolutionLayerPaddingTest", "[ANNLayerTest]")
...
+  // Output-Size should be 2 x 2.
+  // Square output.
+  LpPooling<> module3(2, 2, 2, 2, 2);
+  module3.InputHeight() = 4;
+  module3.InputWidth() = 4;
+  module3.Forward(input, output);
+  // Calculated using torch.nn.LPPool2d().
+  REQUIRE(arma::accu(output) - 77.0 == Approx(0.0).margin(2e-5));
+  REQUIRE(output.n_elem == 4);
+
```
Comment (zoq) [inline src/mlpack/tests/ann_layer_test.cpp:3884]: Looks like we can remove the extra line here.

File: src/mlpack/tests/ann_layer_test.cpp [hunk 5]
```diff
@@ -3838,6 +3838,54 @@ TEST_CASE("TransposedConvolutionLayerPaddingTest", "[ANNLayerTest]")
...
+}
+
+
+
```
Comment (zoq) [inline src/mlpack/tests/ann_layer_test.cpp:3888]:
```suggestion

```

Removed some extra lines here.

**Comment by abh2k** [comment]
shall I    include cmath in `lp_pooling.hpp`  to use cmath?

**Comment by abh2k** [comment]
/Desktop/mlpack/src/mlpack/methods/ann/layer/lp_pooling.hpp:117:11: error: ‘size_t& mlpack::ann::LpPooling<InputDataType, OutputDataType>::NormType() const’ cannot be overloaded with ‘size_t mlpack::ann::LpPooling<InputDataType, OutputDataType>::NormType() const’
  117 |   size_t& NormType() const { return norm_type; }

I am getting this error. But a       similar declaration of  kernalWidth is also done.

**Comment by abh2k** [comment]
@zoq  the build is failing and i am not able to figure it out.

File: src/mlpack/methods/ann/layer/layer_types.hpp [hunk 6]
```diff
@@ -282,6 +283,7 @@ using LayerTypes = boost::variant<
     LSTM<arma::mat, arma::mat>*,
     MaxPooling<arma::mat, arma::mat>*,
     MeanPooling<arma::mat, arma::mat>*,
+    LpPooling<arma::mat, arma::mat>*,
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/layer_types.hpp:286]: Can you add the layer to `MoreTypes` (same file) instead and see if that works, we might run into some limitations here.

Comment (abh2k) [reply src/mlpack/methods/ann/layer/layer_types.hpp]: I think it worked.

Comment (zoq) [reply src/mlpack/methods/ann/layer/layer_types.hpp]: Okay, great, once https://github.com/mlpack/mlpack/pull/2777 is ready this is going to be a lot easier.

**Review by zoq — APPROVED** [review]
Looks good to me, for future contributions it would be useful to use more descriptive commit messages, I know it's not fun but incredibly useful for someone to figure out what changed between commits.

File: src/mlpack/methods/ann/layer/lp_pooling.hpp [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+    {
+      for (size_t i = 0, rowidx = 0; i < output.n_rows;
+           ++i, rowidx += strideWidth)
+      {
+        arma::mat subInput = input(
+            arma::span(rowidx, rowidx + kernelWidth - 1 - offset),
+            arma::span(colidx, colidx + kernelHeight - 1 - offset));
+
+        output(i, j) = pow(arma::accu(arma::pow(subInput,
+            normType)), 1.0/normType);
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/lp_pooling.hpp:175]:
```suggestion
            normType)), 1.0 / normType);
```

Hey, small style fix. We generally use space between operands and operators. Thanks.

File: src/mlpack/methods/ann/layer/lp_pooling.hpp [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+
+    arma::Mat<eT> unpooledError;
+    for (size_t j = 0; j < input.n_cols - cStep; j += cStep)
+    {
+      for (size_t i = 0; i < input.n_rows - rStep; i += rStep)
+      {
+        const arma::Mat<eT>& inputArea = input(arma::span(i, i + rStep - 1),
+            arma::span(j, j + cStep - 1));
+        size_t sum = pow(arma::accu(arma::pow(inputArea, normType)),
+            (normType-1) / normType);
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/lp_pooling.hpp:202]:
```suggestion
            (normType - 1) / normType);
```
Same as above. Thanks.

**Review by kartikdutt18 — COMMENTED** [review]
Hey @abh2k, Thanks for taking the time to contribute this PR. This looks good to me. I have mentioned minor style fixes.
Regards.

**Comment by zoq** [comment]
Thanks again for the contribution.

**Pull request merged**